### PR TITLE
Add Item No Message Passing GCMC

### PR DESCRIPTION
**Pull request by mski-iksm**
Added `GraphNoHiddenConvolutionalMatrixCompletionGraph` , a model making item embedding without using item hidden layer. (User embedding is generated as `GraphConvolutionalMatrixCompletionGraph` .)

### CODE REVIEW

File: redshells/model/graph_convolutional_matrix_completion.py [hunk 1]
```diff
@@ -152,17 +99,32 @@ def _adjust_adjacency_matrix(self, normalization_type):
...
+    def _add_user_feature(self, encoder_hidden_size, encoder_size, side_initial="glorot_normal"):
+        if len(self.input_user_features) != 0:
+            self.user_feature_layers = [self._feature_convert_layer(encoder_hidden_size, kernel_initializer=layer_init) for layer_init in self.gcmc_user_feature_layers_init]
+            x = tf.reduce_prod([layer(feature)+1.0 for layer, feature in zip(self.user_feature_layers, self.input_user_features)], axis=0)
+            self.user_side_info_layer = tf.keras.layers.Dense(encoder_size, use_bias=False, activation='relu', kernel_initializer=side_initial)
+            x = tf.gather(self.user_side_info_layer(x), self.input_user_feature_indices)
+            if self.user_encoder is None:
+                self.user_encoder = x
+            else:
+                self.user_encoder += x  # TODO: discussion for simple addition
```
Comment (Teppei-Kanayama) [inline redshells/model/graph_convolutional_matrix_completion.py]: @mski-iksm 
Is it OK for 'TODO' to be still remaining?

File: redshells/model/graph_convolutional_matrix_completion.py [hunk 2]
```diff
@@ -226,7 +310,7 @@ def __init__(self,
                  normalization_type: str,
                  weight_sharing: bool = True,
                  ignore_item_embedding: bool = False,
-                 save_directory_path: str = None) -> None:
+                 save_directory_path: str = None):
```
Comment (Teppei-Kanayama) [inline redshells/model/graph_convolutional_matrix_completion.py]: @mski-iksm 
Why did you delete this type hinting?

File: redshells/model/graph_convolutional_matrix_completion.py [hunk 3]
```diff
@@ -408,21 +528,53 @@ def _make_graph(self) -> GraphConvolutionalMatrixCompletionGraph:
             ignore_item_embedding=self.ignore_item_embedding)
 
     @staticmethod
-    def _eliminate(matrix: sp.csr_matrix, user_indices, item_indices):
-        matrix = matrix.copy()
-        # `lil_matrix` is too slow
-        matrix[list(user_indices), list(item_indices)] = 0
-        matrix.eliminate_zeros()
-        return matrix
+    def load(file_path: str) -> str:
```
Comment (Teppei-Kanayama) [inline redshells/model/graph_convolutional_matrix_completion.py]: It seems this function doesn't return `str`.

File: redshells/model/graph_convolutional_matrix_completion.py [hunk 4]
```diff
@@ -213,7 +175,129 @@ def _to_constant(x):
         return tf.constant(x) if x is not None else None
 
 
-class GraphConvolutionalMatrixCompletion(object):
+class GraphConvolutionalMatrixCompletionGraph(GraphConvolutionalMatrixCompletionGraphCore):
```
Comment (hirosassa) [inline redshells/model/graph_convolutional_matrix_completion.py]: This class name seems to be too long. It may be a good chance to reconsideration for the class architecture.

File: redshells/model/graph_convolutional_matrix_completion.py (hunk 1 above)
Comment (mski-iksm) [reply redshells/model/graph_convolutional_matrix_completion.py]: I've resolved the TODOs.

File: redshells/model/graph_convolutional_matrix_completion.py (hunk 2 above)
Comment (mski-iksm) [reply redshells/model/graph_convolutional_matrix_completion.py]: It's a mistake. I've fixed it.

File: redshells/model/graph_convolutional_matrix_completion.py (hunk 3 above)
Comment (mski-iksm) [reply redshells/model/graph_convolutional_matrix_completion.py]: I've fixed the type hinting.

File: redshells/model/graph_convolutional_matrix_completion.py (hunk 4 above)
Comment (mski-iksm) [reply redshells/model/graph_convolutional_matrix_completion.py]: I've refactored the Graph class architecture.

File: redshells/model/graph_convolutional_matrix_completion.py [hunk 5]
```diff
@@ -302,49 +395,46 @@ def get_item_feature(self, user_ids: List, item_ids: List, additional_dataset: G
...
         rating_adjacency_matrix = dataset.train_rating_adjacency_matrix()
         user_indices, item_indices = dataset.to_indices(user_ids, item_ids)
         if not with_user_embedding:
-            user_indices = np.array([0] * len(user_indices))  # TODO use default user index.
+            user_indices = np.array([0] * len(user_indices))
 
         user_feature_indices, item_feature_indices = dataset.to_feature_indices(user_ids, item_ids)
         input_data = dict(user=user_indices, item=item_indices, user_feature_indices=user_feature_indices, item_feature_indices=item_feature_indices)
-        feed_dict = cls._feed_dict(input_data, graph, dataset, rating_adjacency_matrix, is_train=False)
+        feed_dict = self._feed_dict(input_data, graph, dataset, rating_adjacency_matrix, is_train=False)
```
Comment (Teppei-Kanayama) [inline redshells/model/graph_convolutional_matrix_completion.py]: @mski-iksm 
It might be better to leave `_predict()` function to be class method, because class method prevents the `_predict()` function from having access to the class attributes and this might lead your code more readable.

File: redshells/model/graph_convolutional_matrix_completion.py [hunk 6]
```diff
@@ -302,49 +395,46 @@ def get_item_feature(self, user_ids: List, item_ids: List, additional_dataset: G
...
         user_indices, item_indices = dataset.to_indices(user_ids, item_ids)
         if not with_user_embedding:
-            user_indices = np.array([0] * len(user_indices))  # TODO use default user index.
+            user_indices = np.array([0] * len(user_indices))
 
         user_feature_indices, item_feature_indices = dataset.to_feature_indices(user_ids, item_ids)
         input_data = dict(user=user_indices, item=item_indices, user_feature_indices=user_feature_indices,
                           item_feature_indices=item_feature_indices)
-        feed_dict = cls._feed_dict(input_data, graph, dataset, rating_adjacency_matrix, is_train=False)
+        feed_dict = self._feed_dict(input_data, graph, dataset, rating_adjacency_matrix, is_train=False)
```
Comment (Teppei-Kanayama) [inline redshells/model/graph_convolutional_matrix_completion.py]: @mski-iksm 
The same comment as above.

File: redshells/model/graph_convolutional_matrix_completion.py [hunk 7]
```diff
@@ -27,163 +28,132 @@ def _convert_sparse_matrix_to_sparse_tensor(x):
...
-            self._add_item_feature(encoder_hidden_size, encoder_size, ignore_item_embedding)
-
-            # decoder
-            self.output = self._decoder(encoder_size=encoder_size, n_rating=n_rating, user_encoder=self.user_encoder, item_encoder=self.item_encoder)
-
-            # output
-            self.probability = tf.nn.softmax(self.output)
-            self.expectation = tf.matmul(self.probability, tf.reshape(self.rating, shape=(-1, 1)))
-            self.rmse = tf.sqrt(tf.reduce_mean(tf.math.square(self.expectation - tf.reshape(tf.to_float(self.input_rating), shape=(-1, 1)))))
+class GraphABC(ABC):
```
Comment (Teppei-Kanayama) [inline redshells/model/graph_convolutional_matrix_completion.py]: @mski-iksm 
In my opinion, `AbstractGraph` sounds better than `GraphABC` in terms of the name of the class because abbreviations should be avoided if possible.

File: redshells/model/graph_convolutional_matrix_completion.py (hunk 5 above)
Comment (mski-iksm) [reply redshells/model/graph_convolutional_matrix_completion.py]: I've changed `_predict()` to classmethod and `_feed_dict()` to staticmethod

File: redshells/model/graph_convolutional_matrix_completion.py (hunk 7 above)
Comment (mski-iksm) [reply redshells/model/graph_convolutional_matrix_completion.py]: I've changed the class name to AbstractGraph.

**Comment by mski-iksm** [comment]
This master PR is closed until a better model is developed.
